lib/ecto/adapters/sql/sandbox.ex

```diff
@@ -1,290 +1,227 @@
+  any other pool. The only difference is that connections are not
+  checked back in automatically but by explicitly calling `checkin/2`.
 
-  defmodule Result do
-    defstruct [:value]
-  end
+  The `mode/2` function can be used to change the pool mode to
+  `:manual`. In this case, each connection must be explicitly
+  checked out before use. This is useful when paired with
+  `checkout/2` which by default wraps the connection in a transaction.
+  This means developers have a safe mechanism for running concurrent
+  tests against the database.
 
-  @behaviour DBConnection
-  @behaviour DBConnection.Pool
+  ## Example
 
-  defstruct [:mod, :state, :status, :adapter]
+  The first step is to configure your database to use the
+  `Ecto.Adapters.SQL.Sandbox` pool. You set those options in your
+  `config/config.exs`:
 
-  @doc false
-  def mode(pool) do
-    DBConnection.execute!(pool, %Query{request: :mode}, [], [pool: __MODULE__])
-  end
+      config :my_app, Repo,
+        pool: Ecto.Adapters.SQL.Sandbox
 
-  @doc false
-  def start_link(mod, opts) do
-    DBConnection.Connection.start_link(__MODULE__, opts(mod, opts))
-  end
+  Since you don't want those options in your production database, we
+  typically recommend to create a `config/test.exs` and add the
+  following to the bottom of your `config/config.exs` file:
 
-  @doc false
-  def child_spec(mod, opts, child_opts \\ []) do
-    DBConnection.Connection.child_spec(__MODULE__, opts(mod, opts), child_opts)
-  end
+      import_config "config/#{Mix.env}.exs"
 
-  @doc false
-  def checkout(pool, opts) do
-    DBConnection.Connection.checkout(pool, opts(opts))
-  end
+  Now with the test database properly configured, you can write
+  transactional tests:
 
-  @doc false
-  def checkin(pool, state, opts) do
-    DBConnection.Connection.checkin(pool, state, opts(opts))
-  end
+      # At the end of your test_helper.exs
+      # Set the pool mode to manual for explicitly checkouts
+      Ecto.Adapters.SQL.Sandbox.mode(TestRepo, :manual)
 
-  @doc false
-  def disconnect(pool, err, state, opts) do
-    DBConnection.Connection.disconnect(pool, err, state, opts(opts))
-  end
+      defmodule PostTest do
+        # Once the model is manual, tests can also be async
+        use ExUnit.Case, async: true
 
-  @doc false
-  def stop(pool, reason, state, opts) do
-    DBConnection.Connection.stop(pool, reason, state, opts(opts))
-  end
-
-  @doc false
-  def connect(opts) do
-    mod     = Keyword.fetch!(opts, :sandbox)
-    adapter = Module.concat(Keyword.fetch!(opts, :adapter), Connection)
-    case apply(mod, :connect, [opts]) do
-      {:ok, state} ->
-        s = %__MODULE__{mod: mod, state: state, status: :run, adapter: adapter}
-        {:ok, s}
-      {:error, _} = error ->
-        error
-    end
-  end
+        setup do
+          # Explicitly get a connection before each test
+          :ok = Ecto.Adapters.SQL.Sandbox.checkout(TestRepo)
```

Comment:
Don't you need to have a teardown block as well, returning the connection? Or is that handled automatically?